### PR TITLE
Check value is not None before inserting into forward index

### PR DESCRIPTION
**Pull request by vangheem**
This fixes zope.index in ZODB 4

### CODE REVIEW

File: src/zope/index/field/index.py [hunk 1]
```diff
@@ -61,12 +61,13 @@ def index_doc(self, docid, value):
...
-            set = self.family.IF.TreeSet()
-            self._fwd_index[value] = set
-        set.insert(docid)
+        if value is not None:
+                # Insert into forward index.
+                set = self._fwd_index.get(value)
+                if set is None:
+                    set = self.family.IF.TreeSet()
+                    self._fwd_index[value] = set
+                set.insert(docid)
```
Comment (pbauer) [inline src/zope/index/field/index.py]: 8 space-indentation?

**Comment by tseaver** [comment]
Thanks for the patch!  A couple of issues:

- This change suppresses an exception for such values on Py3k:  are we sure we want to do that, rather than raising the exception even on Python2?
- Whichever way that choice goes, the "value is None" case needs a new test, with appropriate assertions (either of the unchanged state of the forward index, or the always-raised exception).

**Comment by vangheem** [comment]
@tseaver I'm not sure this has anything to do with p2k vs p3k. In the latest BTree, which ZODB depends one, you can not have None values as keys.

**Comment by tseaver** [comment]
Ah, right -- I forgot that bit.

As it stands, this PR means that we won't be able to search for documents having `None` as their value, although they will be present in the reverse index.

**Comment by vangheem** [comment]
I don't really know enough about how zope.index works so this is probably short sighted. I had thought the rev_index was where the search was done on.

Do we catch the TypeError and say that None values aren't allowed with ZODB 4? 

I also don't know what happens for upgraded ZODBs where there are existing None values.

**Comment by tseaver** [comment]
Thanks again!

**Pull request merged**